docs/src/design/decoder/constraints.md

```diff
@@ -0,0 +1,578 @@
+
+When executing `SPAN` or `RESPAN` operation, the next value in $sp$ column must be set to $1$:
+
+> $$
+(f_{span} + f_{respan}) \cdot (1 - sp') = 0 \text{ | degree } = 6
+$$
+
+When the next operation is `END` or `RESPAN`, the next value in $sp$ column must be set $0$.
+
+> $$
+(f_{end}' + f_{respan}') \cdot sp' = 0 \text{ | degree } = 6
+$$
+
+In all other cases, the value in $sp$ column must be copied over to the next row:
+
+> $$
+(1 - f_{span} - f_{respan} - f_{end}' - f_{respan}') \cdot (sp' - sp) = 0 \text{ | degree } = 6
+$$
+
+Additionally, we will need to impose a boundary constraint which specifies that the first value in $sp = 0$. Note, however, that we do not need to impose a constraint ensuring that values in $sp$ are binary - this will follow naturally from the above constraints.
+
+Also, note that the combination of the above constraints makes it impossible to execute `END` or `RESPAN` operations right after `SPAN` or `RESPAN` operations.
+
+### Block address constraints
+When we are inside a *span* block, values in block address columns (denoted as $a$) must remain the same. This can be enforced with the following constraint:
+
+> $$
+sp \cdot (a' - a) = 0 \text{ | degree } = 2
+$$
+
+Notice that this constraint does not apply when we execute any of the control flow operations. For such operations, the prover sets the value of the $a$ column non-deterministically, except for the `RESPAN` operation. For the `RESPAN` operation the value in the $a$ column is incremented by $8$, which is enforced by a constraint described previously.
+
+Notice also that this constraint implies that when the next operation is the `END` operation, the value in the $a$ colum must also be copied over to the next row. This is exactly the behavior we want to enforce so that when the `END` operation is executed, the block address is set to the address of the current span batch.
+
+### Group count constraints
+The `group_count` column (denoted as $gc$) is used to keep track of the number of operation groups which remains to be executed in a span block.
+
+In the beginning of a span block (i.e., when `SPAN` operation is executed), the prover sets the value of $gc$ non-deterministically. This value is subsequently decremented according to the rules described below. By the time we exit the span block (i.e., when `END` operation is executed), the value in $gc$ must be $0$.
+
+The rules for decrementing values in the $gc$ column are as follows:
+* The count cannot be decremented by more than $1$ in a single row.
+* When an operation group is fully executed (which happens when $h_0 = 0$ inside a span block), the count is decremented by $1$.
+* When `PUSH` operation is executed, the count is decremented by $1$. These 3 rules together mean that `PUSH` operation cannot be the last operation in an operation group (otherwise the count would have to be decremented by $2$).
+* When `SPAN` or `RESPAN` operation is executed, the value is decremented by $1$.
```

Comment:
Good point! Updated. One thing to note: `SPAN` and `RESPAN` operations are never included in operation groups because operation groups contain only user operations.